etp-backend/src/main/clj/solita/common/smtp.clj

```diff
@@ -0,0 +1,62 @@
+(ns solita.common.smtp
+  (:require [clojure.tools.logging :as log])
+  (:import (javax.mail Message Message$RecipientType Session)
+           (javax.mail.internet InternetAddress
+                                MimeMessage
+                                MimeBodyPart
+                                MimeMultipart)))
+
+(defn mail-properties [port]
+  (doto (System/getProperties)
+    (.put "mail.transport.protocol" "smtp")
+    (.put "mail.smtp.port" port)
+    (.put "mail.smtp.starttls.enable" "true")
+    (.put "mail.smtp.auth" "true")))
+
+(defn session [properties]
+  (Session/getDefaultInstance properties))
+
+(defn body-mime-body-part [body content-type]
+  (doto (MimeBodyPart.)
+    (.setContent body content-type)))
+
+(defn attachment-mime-body-part [attachment]
+  (doto (MimeBodyPart.)
+    (.attachFile attachment)))
+
+(defn multipart [& mime-body-parts]
+  (let [mime-multi-part (MimeMultipart.)]
+    (doseq [mime-body-part mime-body-parts]
+      (.addBodyPart mime-multi-part mime-body-part))
+    mime-multi-part))
+
+(defn mime-message [session from-email from-name to subject multipart]
+  (doto (MimeMessage. session)
+    (.setFrom (InternetAddress. from-email from-name))
+    (.setRecipient Message$RecipientType/TO (InternetAddress. to))
+    (.setSubject subject)
+    (.setContent multipart)))
+
+(defn transport [session]
+  (.getTransport session))
+
+(defn send-email! [host port username password from-email from-name to subject
+                   body content-type attachments]
+  (let [properties (mail-properties port)
+        session (session properties)
+        body-mime-body-part (body-mime-body-part body content-type)
+        attachments-mime-body-part (map attachment-mime-body-part
+                                        attachments)
+        multipart (apply multipart body-mime-body-part attachments-mime-body-part)
+        mime-message (mime-message session
+                                   from-email
+                                   from-name
+                                   to
+                                   subject
+                                   multipart)]
+    (doto (transport session)
+      (.connect host username password)
+      (.sendMessage mime-message (.getAllRecipients mime-message))
```

Comment:
I am not very familiar with `doto`, but it seems like it does not guarantee that the session gets closed. Should it be enforced?

Reply:
Good question. Transport is not Closeable, but maybe the sendMessage -part should be in try-catch-finally.